zkp/golang/pkg/core/storage.go

```diff
@@ -31,3 +33,41 @@ type Storage interface {
 	// Close closes the storage resource
 	Close()
 }
+
+const (
+	// we use a table to store the root node indexes for
+	// all the merkle trees in the database
+	TreeRootsTable = "merkelTreeRoots"
+	// we use a separate table to store the nodes of each
+	// sparse merkle tree by using the following name as
+	// the prefix, followed by the name of the tree
+	NodesTablePrefix = "smtNodes_"
+)
+
+// SqlDBProvider is the interface for providing access to a SQL database to
+// the storage layer implementations that are backed by a SQL database.
+type SqlDBProvider interface {
+	DB() *gorm.DB
+	Close()
+}
+
+// SMTRoot is used to persist tree root in SQL databases via gorm
+type SMTRoot struct {
+	// the name of the merkle tree
+	Name string `gorm:"primaryKey"`
+	// this must be the hex bytes of the root index
+	// following the big-endian encoding
+	RootIndex string `gorm:"type:text"`
+}
+
+// SMTNode is the structure of a node in the merkle tree.
+// It only captures the reference key and the index of the node.
+// The value properties of a node are local states that are
+// handled outside of the merkle tree library.
+type SMTNode struct {
+	RefKey     string `gorm:"primaryKey"`
+	Type       byte
+	Index      *string // only leaf nodes have an index
+	LeftChild  *string // only branch nodes have children
+	RightChild *string
```

Comment:
do you need to annotate this with `gorm` of it will default to VARCHAR for strings?

Reply:
What is default length that GROM generated for VARCHAR? For example SQLITE VARCHAR is unbounded but PSQL is bounder

Reply:
looks like there's a way to specify a size in the gorm annotation. so added that to the structs for both tables